### PR TITLE
[Diagnostics] [Typechecker] Emit fix-its for witness mismatches

### PR DESCRIPTION
**Pull request by theblixguy**
Saw a couple of FIXMEs, so I decided to address them and then also add fix-its for the rest of the cases.

This PR adds fix-its for when the witness does not match the requirements.

1. I didn't add a fix-it for the `@nonobjc` case because the attribute's location does not look correct when I pass it to `fixItReplace` (i.e. the replaced range does not look correct).
2. I couldn't reproduce two diagnostics - one for `nonmutating` and one for `__consuming`.

### CODE REVIEW

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -2072,9 +2072,14 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
                    (unsigned)match.MissingRequirement->getKind());
     break;
 
-  case MatchKind::ThrowsConflict:
-    diags.diagnose(match.Witness, diag::protocol_witness_throws_conflict);
+  case MatchKind::ThrowsConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(witness, diag::protocol_witness_throws_conflict);
+    assert(isa<FuncDecl>(witness));
+    auto FD = cast<FuncDecl>(witness);
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: The assertion is implied by the use of `cast` instead of `dyn_cast`.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 2]
```diff
@@ -2072,9 +2072,14 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
+    diag.fixItRemove(FD->getThrowsLoc());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: I'm not sure this is a good fix-it to offer. `throws` is unlikely to be accidentally added, and if there's anything relying on it in the body of the function we'd just get into an error loop.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 3]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-  case MatchKind::StaticNonStaticConflict:
-    // FIXME: Could emit a Fix-It here.
-    diags.diagnose(match.Witness, diag::protocol_witness_static_conflict,
-                   !req->isInstanceMember());
+  case MatchKind::StaticNonStaticConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(witness, diag::protocol_witness_static_conflict,
+                               !req->isInstanceMember());
+    if (req->isInstanceMember()) {
+      assert(isa<FuncDecl>(witness) || isa<VarDecl>(witness));
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Subscripts?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 4]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
+      auto loc = isa<FuncDecl>(witness)
+                     ? cast<FuncDecl>(witness)->getStaticLoc()
+                     : cast<VarDecl>(witness)
+                           ->getParentPatternBinding()
+                           ->getStaticLoc();
+      diag.fixItRemove(loc);
+    } else {
+      diag.fixItInsert(witness->getAttributeInsertionLoc(true),
+                       getTokenText(tok::kw_static));
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: For comparison, this one seems more valid: you can *forget* to write `static` when you mean it, and converting a static member to a non-static one is less likely to result in errors (unless it's already in use).

File: lib/Sema/TypeCheckProtocol.cpp [hunk 5]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-                   diag::protocol_witness_prefix_postfix_conflict, false,
-                   match.Witness->getAttrs().hasAttribute<PostfixAttr>() ? 2
-                                                                         : 0);
+  case MatchKind::PrefixNonPrefixConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(
+        witness, diag::protocol_witness_prefix_postfix_conflict, false,
+        witness->getAttrs().hasAttribute<PostfixAttr>() ? 2 : 0);
+    if (witness->getAttrs().hasAttribute<PostfixAttr>()) {
+      auto attr = witness->getAttrs().getAttribute<PostfixAttr>();
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Nitpick: You can collapse this to `if (auto attr = …)`

File: lib/Sema/TypeCheckProtocol.cpp [hunk 6]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-  case MatchKind::SettableConflict:
-    diags.diagnose(match.Witness, diag::protocol_witness_settable_conflict);
+  }
+
+  case MatchKind::SettableConflict: {
+    auto witness = match.Witness;
+    auto diag =
+        diags.diagnose(witness, diag::protocol_witness_settable_conflict);
+    auto PBD = cast<VarDecl>(witness)->getParentPatternBinding();
+    diag.fixItReplace(PBD->getStartLoc(), getTokenText(tok::kw_var));
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: This isn't necessarily correct; `isSettable` also checks read-only computed properties.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 7]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-                                                                         : 0);
+  case MatchKind::PrefixNonPrefixConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(
+        witness, diag::protocol_witness_prefix_postfix_conflict, false,
+        witness->getAttrs().hasAttribute<PostfixAttr>() ? 2 : 0);
+    if (witness->getAttrs().hasAttribute<PostfixAttr>()) {
+      auto attr = witness->getAttrs().getAttribute<PostfixAttr>();
+      diag.fixItReplace(attr->getLocation(), "prefix");
+    }
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Else insert it, maybe?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 8]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-                   diag::protocol_witness_mutation_modifier_conflict,
-                   SelfAccessKind::Mutating);
+  }
+  case MatchKind::MutatingConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(
+        witness, diag::protocol_witness_mutation_modifier_conflict,
+        SelfAccessKind::Mutating);
+    auto attr = witness->getAttrs().getAttribute<MutatingAttr>();
+    diag.fixItRemove(attr->getLocation());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Questionable, again.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 9]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-                   diag::protocol_witness_mutation_modifier_conflict,
-                   SelfAccessKind::NonMutating);
+  }
+  case MatchKind::NonMutatingConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(
+        witness, diag::protocol_witness_mutation_modifier_conflict,
+        SelfAccessKind::NonMutating);
+    auto attr = witness->getAttrs().getAttribute<NonMutatingAttr>();
+    diag.fixItRemove(attr->getLocation());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: *Definitely* questionable even though it's less likely to cause compilation errors. In practice, though, this conflict kind is unused because a non-mutating witness *can* satisfy a mutating requirement. It can be removed in a separate commit.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 10]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-                   diag::protocol_witness_mutation_modifier_conflict,
-                   SelfAccessKind::Consuming);
+  }
+  case MatchKind::ConsumingConflict: {
+    auto witness = match.Witness;
+    auto diag = diags.diagnose(
+        witness, diag::protocol_witness_mutation_modifier_conflict,
+        SelfAccessKind::Consuming);
+    auto attr = witness->getAttrs().getAttribute<ConsumingAttr>();
+    diag.fixItRemove(attr->getLocation());
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Questionable again.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 11]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
-  case MatchKind::RethrowsConflict:
-    // FIXME: Could emit a Fix-It here.
-    diags.diagnose(match.Witness, diag::protocol_witness_rethrows_conflict);
+  }
+  case MatchKind::RethrowsConflict: {
+    auto witness = match.Witness;
+    auto diag =
+        diags.diagnose(match.Witness, diag::protocol_witness_rethrows_conflict);
+    assert(isa<FuncDecl>(witness));
+    auto FD = cast<FuncDecl>(witness);
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: Redundant assert again.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 12]
```diff
@@ -2089,53 +2094,95 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
+    diag.fixItReplace(FD->getThrowsLoc(), getTokenText(tok::kw_rethrows));
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckProtocol.cpp]: This does seem a little more probable: someone doesn't know about `rethrows` yet but they happen to have implemented their function that way.

**Comment by jrose-apple** [comment]
Forgot to say: thanks for picking these up!

File: lib/Sema/TypeCheckProtocol.cpp (hunk 3 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: Oops.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 7 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: I did that, but later removed because I noticed we already suggest a fix-it to add that (`{prefix|postfix} unary operator missing '{prefix|postfix}' modifier`).

File: lib/Sema/TypeCheckProtocol.cpp (hunk 10 above)
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: Tbh I just added them because they all had a FIXME, but I also couldn't reproduce it. I have removed it now.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 7 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckProtocol.cpp]: Maybe leave a comment?

Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: Done!

**Comment by theblixguy** [comment]
@swift-ci please smoke test

**Comment by theblixguy** [comment]
@swift-ci please smoke test

File: test/decl/protocol/req/witness_fix_its.swift [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+  static prefix func ^^^(value: Self) -> Int // expected-note {{protocol requires function '^^^' with type '(ConformsToFoo) -> Int'; do you want to add a stub?}}
+  static postfix func ^^^^(value: Self) -> Int // expected-note {{protocol requires function '^^^^' with type '(ConformsToFoo) -> Int'; do you want to add a stub?}}
+  func bar4(closure: () throws -> Int) rethrows // expected-note {{protocol requires function 'bar4(closure:)' with type '(() throws -> Int) throws -> ()'; do you want to add a stub?}}
+  var bar5: Int { get set } // expected-note {{protocol requires property 'bar5' with type 'Int'; do you want to add a stub?}}
+  static subscript(_ pos: Int) -> Int { get } // expected-note {{protocol requires subscript with type '(Int) -> Int'; do you want to add a stub?}}
+}
+
+struct ConformsToFoo: Foo { // expected-error {{type 'ConformsToFoo' does not conform to protocol 'Foo'}}
+  let bar1: Int // expected-note {{candidate is not settable, but protocol requires it}}{{3-6=var}}
+  var bar2: Int // expected-note {{candidate operates on an instance, not a type as required}}{{3-3=static}}
```
Comment (jrose-apple) [inline test/decl/protocol/req/witness_fix_its.swift]: This needs a trailing space, yeah?

File: lib/Sema/TypeCheckProtocol.cpp [hunk 14]
```diff
@@ -2089,53 +2089,89 @@ diagnoseMatch(ModuleDecl *module, NormalProtocolConformance *conformance,
...
+    if (req->isInstanceMember()) {
+      SourceLoc loc;
+      if (auto FD = dyn_cast<FuncDecl>(witness)) {
+        loc = FD->getStaticLoc();
+      } else if (auto VD = dyn_cast<VarDecl>(witness)) {
+        loc = VD->getParentPatternBinding()->getStaticLoc();
+      } else if (auto SD = dyn_cast<SubscriptDecl>(witness)) {
+        loc = SD->getStaticLoc();
+      } else {
+        llvm_unreachable("Unexpected witness");
```
[the remark this answers is not in the document]
Comment (theblixguy) [reply lib/Sema/TypeCheckProtocol.cpp]: > rely on plain getStaticLoc() for anything else

What do you mean by that? Because I can't access `getStaticLoc()` until I cast it the witness to one of these specific decls.

> did we ever implement the thing where enum elements can satisfy requirements?

Nope, enum cases can't satisfy protocol requirements at the moment (see SR-3170).

File: test/decl/protocol/req/witness_fix_its.swift (hunk 13 above)
Comment (theblixguy) [reply test/decl/protocol/req/witness_fix_its.swift]: You're right! I have fixed it now.

File: lib/Sema/TypeCheckProtocol.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (jrose-apple) [reply lib/Sema/TypeCheckProtocol.cpp]: Oh, whoops. I thought `getStaticLoc` was on Decl or ValueDecl. Withdrawn.

File: test/decl/protocol/req/witness_fix_its.swift (hunk 13 above)
Comment (jrose-apple) [reply test/decl/protocol/req/witness_fix_its.swift]: …and now your tests need updating.

**Review by jrose-apple — APPROVED** [review]

**Comment by theblixguy** [comment]
@swift-ci please smoke test
